dashboard/app/assets/javascripts/header.js.erb

```diff
@@ -2,110 +2,575 @@
+      if (state.showStageLinks) {
+        el = $('<span>').addClass('header_popup_link');
+
+        if (state.popped) {
+          $('<div>').addClass('header_popup_link_glyph').html("&#x25B2;").appendTo(el);
+          $('<div>').addClass('header_popup_link_text').text('<%= I18n.t(:less) %>').appendTo(el);
+        } else {
+          $('<div>').addClass('header_popup_link_glyph').html("&#x25BC;").appendTo(el);
+          $('<div>').addClass('header_popup_link_text').text('<%= I18n.t(:more) %>').appendTo(el);
+        }
+        el.appendTo($container);
+
+        $('.freeplay_links').show();
+      } else {
+        $('.freeplay_links').hide();
+      }
+
+      // Popup stage navigation
+      if (state.popped) {
+        var header_popup = $('<div>').addClass('header_popup').appendTo($middle);
+
+        el = $('<div>').addClass('header_popup_header').appendTo(header_popup);
+        $('<span>').text('<%= I18n.t("nav.popup.progress") %>').appendTo(el);
+        if (state.linesOfCodeText)
+          $('<div>').addClass('header_text').css('float','right').text(state.linesOfCodeText).appendTo(el);
+
+        el = $('<div>').addClass('header_popup_body').appendTo(header_popup);
+        el.html(state.stageHtml || '<div class="loading"></div>');
+
+        el = $('<div>').addClass('header_popup_footer').appendTo(header_popup);
+        $('<div>').addClass('header_popup_close').text('<%= I18n.t("nav.popup.close") %>').appendTo(el);
+
+        // Catch clicks anywhere else and close the popup
+        $(document).on('click', this.onModalClick);
+      } else {
+        $(document).off('click', this.onModalClick);
+      }
+    }
+
+    HeaderProgress.prototype.componentDidUpdate = function() {
+      // This is totally cheating by React rules.  But it's intentional: we don't want to force another
+      // update so we just edit the state directly.  (React would notice that the change had no DOM effect,
+      // but we don't have that optimization.)
+      if (this.state.stageHtml && this.state.jumpToTrophies) {
+        var off = $('#trophies').offset();
+        if (off)
+          window.scrollTo(0, +off.top);
+        delete this.state.jumpToTrophies;
+      }
+    }
+
+    // Three possible arguments: true, false, or "trophies" (the last one autoscrolls the page to the bottom of the popup)
+    HeaderProgress.prototype.showPopup = function(show) {
+      this.setState({
+        popped: !!show,
+        jumpToTrophies: show == 'trophies'
+      });
+
+      // Have we loaded the stage HTML already?  If not, load it now.
+      if (this.state.popped && (this.state.stageHtml === undefined)) {
+        this.state.stageHtml = false; // prevent multiple loads
+
+        // Ask the script store to load a particular script.
+        this.scriptStore.loadHTML({
+          script_id: this.state.scriptId,
+          script_level_id: this.state.scriptLevelId
+        });
+
+      }
+    }
+
+    HeaderProgress.prototype.onClick = function(ev) {
```

Comment:
Click handling in my UIComponent is done with a single el.on('click') on the whole owned region.  I have a single click handler that tests against a number of possibilities to find out what was clicked.

This would be problematic for devs if we keep this format, but we won't.  React doesn't work this way (or rather, it does, but behind the scenes & calls unique functions)